Behavior difference: Assigning `Object.values(obj)` to `obj[field]` gives error on tsgo but not on tsc

Could be related to https://github.com/microsoft/typescript-go/issues/3667#issuecomment-4359947311, since this issue is again about some self-referential assignments.

---

## Steps to reproduce

1. Start a plain new project
    - "@typescript/native-preview": "7.0.0-dev.20260503.1",
    - "typescript": "6.0.3"
2. Run `npx tsgo --init` to get the default tsconfig. Turn on these flags as well.
    - "allowJs": true,
    - "checkJs": true,
    - "stableTypeOrdering": true,
3. Add `src/main.js`. Run `npx tsc --noEmit` & `npx tsgo --noEmit` to see the difference.

```js
// src/main.js
export class C {
    method() {
        const obj = {};
        obj.foo = 'foo';
        obj.bar = 'bar';
        obj.buzz = Object.values(obj);
    }
}
```


## Behavior with `typescript@6.0`

No error at all.

## Behavior with `tsgo`

```
src/main.js:7:9 - error TS7022: 'buzz' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.

7         obj.buzz = Object.values(obj);
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in src/main.js:7
```

--- a/internal/checker/relater.go
+++ b/internal/checker/relater.go
@@ -4583,16 +4583,16 @@ func (r *Relater) typeRelatedToIndexInfo(source *Type, targetInfo *IndexInfo, re
 	return TernaryFalse
 }
 
-/**
- * Return true if type was inferred from an object literal, written as an object type literal, or is the shape of a module
- * with no call or construct signatures.
- */
+// Return true if type was inferred from a non-expando object literal, written as an object type literal, or is the shape of a module
+// with no call or construct signatures.
 func (c *Checker) isObjectTypeWithInferableIndex(t *Type) bool {
 	if t.flags&TypeFlagsIntersection != 0 {
 		return core.Every(t.Types(), c.isObjectTypeWithInferableIndex)
 	}
-	return t.symbol != nil && t.symbol.Flags&(ast.SymbolFlagsObjectLiteral|ast.SymbolFlagsTypeLiteral|ast.SymbolFlagsEnum|ast.SymbolFlagsValueModule) != 0 &&
-		t.symbol.Flags&ast.SymbolFlagsClass == 0 && !c.typeHasCallOrConstructSignatures(t) ||
+	return t.symbol != nil &&
+		(t.symbol.Flags&(ast.SymbolFlagsObjectLiteral|ast.SymbolFlagsTypeLiteral) != 0 && len(t.symbol.Exports) == 0 ||
+			t.symbol.Flags&(ast.SymbolFlagsEnum|ast.SymbolFlagsValueModule) != 0 && t.symbol.Flags&ast.SymbolFlagsClass == 0) &&
+		!c.typeHasCallOrConstructSignatures(t) ||
 		t.objectFlags&ObjectFlagsObjectRestType != 0 ||
 		t.objectFlags&ObjectFlagsReverseMapped != 0 && c.isObjectTypeWithInferableIndex(t.AsReverseMappedType().source)
 }

--- a/testdata/baselines/reference/compiler/expandoNoInferredIndex.symbols
+++ b/testdata/baselines/reference/compiler/expandoNoInferredIndex.symbols
@@ -1,0 +1,33 @@
+//// [tests/cases/compiler/expandoNoInferredIndex.ts] ////
+
+=== main.js ===
+export class C {
+>C : Symbol(C, Decl(main.js, 0, 0))
+
+    method() {
+>method : Symbol(C.method, Decl(main.js, 0, 16))
+
+        const obj = {};
+>obj : Symbol(obj, Decl(main.js, 2, 13))
+
+        obj.foo = 'foo';
+>obj.foo : Symbol(foo, Decl(main.js, 2, 23))
+>obj : Symbol(obj, Decl(main.js, 2, 13))
+>foo : Symbol(foo, Decl(main.js, 2, 23))
+
+        obj.bar = 'bar';
+>obj.bar : Symbol(bar, Decl(main.js, 3, 24))
+>obj : Symbol(obj, Decl(main.js, 2, 13))
+>bar : Symbol(bar, Decl(main.js, 3, 24))
+
+        obj.buzz = Object.values(obj);
+>obj.buzz : Symbol(buzz, Decl(main.js, 4, 24))
+>obj : Symbol(obj, Decl(main.js, 2, 13))
+>buzz : Symbol(buzz, Decl(main.js, 4, 24))
+>Object.values : Symbol(ObjectConstructor.values, Decl(lib.es2017.object.d.ts, --, --), Decl(lib.es2017.object.d.ts, --, --))
+>Object : Symbol(Object, Decl(lib.es5.d.ts, --, --), Decl(lib.es5.d.ts, --, --))
+>values : Symbol(ObjectConstructor.values, Decl(lib.es2017.object.d.ts, --, --), Decl(lib.es2017.object.d.ts, --, --))
+>obj : Symbol(obj, Decl(main.js, 2, 13))
+    }
+}
+

--- a/testdata/baselines/reference/compiler/expandoNoInferredIndex.types
+++ b/testdata/baselines/reference/compiler/expandoNoInferredIndex.types
@@ -1,0 +1,40 @@
+//// [tests/cases/compiler/expandoNoInferredIndex.ts] ////
+
+=== main.js ===
+export class C {
+>C : C
+
+    method() {
+>method : () => void
+
+        const obj = {};
+>obj : { foo: string; bar: string; buzz: any[]; }
+>{} : { foo: string; bar: string; buzz: any[]; }
+
+        obj.foo = 'foo';
+>obj.foo = 'foo' : "foo"
+>obj.foo : string
+>obj : { foo: string; bar: string; buzz: any[]; }
+>foo : string
+>'foo' : "foo"
+
+        obj.bar = 'bar';
+>obj.bar = 'bar' : "bar"
+>obj.bar : string
+>obj : { foo: string; bar: string; buzz: any[]; }
+>bar : string
+>'bar' : "bar"
+
+        obj.buzz = Object.values(obj);
+>obj.buzz = Object.values(obj) : any[]
+>obj.buzz : any[]
+>obj : { foo: string; bar: string; buzz: any[]; }
+>buzz : any[]
+>Object.values(obj) : any[]
+>Object.values : { <T>(o: { [s: string]: T; } | ArrayLike<T>): T[]; (o: {}): any[]; }
+>Object : ObjectConstructor
+>values : { <T>(o: { [s: string]: T; } | ArrayLike<T>): T[]; (o: {}): any[]; }
+>obj : { foo: string; bar: string; buzz: any[]; }
+    }
+}
+

--- a/testdata/tests/cases/compiler/expandoNoInferredIndex.ts
+++ b/testdata/tests/cases/compiler/expandoNoInferredIndex.ts
@@ -1,0 +1,14 @@
+// @checkJs: true
+// @noEmit: true
+
+// https://github.com/microsoft/typescript-go/issues/3691
+
+// @filename: main.js
+export class C {
+    method() {
+        const obj = {};
+        obj.foo = 'foo';
+        obj.bar = 'bar';
+        obj.buzz = Object.values(obj);
+    }
+}

